### PR TITLE
Add CvT

### PR DESCRIPTION
**Pull request by NielsRogge**
# What does this PR do?

Co-authored-by: AnugunjNaman <anugunjjha@gmail.com>

This PR adds CvT (Convolutional Vision Transformer) by Microsoft Research.

I just cleaned up the branch of @AnugunjNaman. 

To do:

- [x] make @AnugunjNaman co-author of this PR

### CODE REVIEW

File: src/transformers/models/cvt/configuration_cvt.py [hunk 1]
```diff
@@ -0,0 +1,148 @@
...
+    defaults will yield a similar configuration to that of the CvT
+    [microsoft/cvt-13](https://huggingface.co/microsoft/cvt-13) architecture.
+
+    Configuration objects inherit from [`PretrainedConfig`] and can be used to control the model outputs. Read the
+    documentation from [`PretrainedConfig`] for more information.
+
+    Args:
+        num_channels (`int`, *optional*, defaults to 3):
+            The number of input channels.
+        patch_sizes (`List[int]`, *optional*, defaults to [7, 3, 3]):
```
Comment (sgugger) [inline src/transformers/models/cvt/configuration_cvt.py:42]: All list defaults should be in backticks (here and all args below).
```suggestion
        patch_sizes (`List[int]`, *optional*, defaults to `[7, 3, 3]`):
```

File: src/transformers/models/cvt/modeling_cvt.py [hunk 2]
```diff
@@ -0,0 +1,753 @@
...
+        config_class=_CONFIG_FOR_DOC,
+        modality="vision",
+        expected_output=_EXPECTED_OUTPUT_SHAPE,
+    )
+    def forward(
+        self,
+        pixel_values=None,
+        output_hidden_states=None,
+        return_dict=None,
+    ):
```
Comment (sgugger) [inline src/transformers/models/cvt/modeling_cvt.py:633]:
```suggestion
    def forward(self, pixel_values=None, output_hidden_states=None, return_dict=None):
```

File: src/transformers/models/cvt/modeling_cvt.py [hunk 3]
```diff
@@ -0,0 +1,753 @@
...
+
+        if not return_dict:
+            output = (logits,) + outputs[2:]
+            return ((loss,) + output) if loss is not None else output
+
+        return ImageClassifierOutput(
+            loss=loss,
+            logits=logits,
+            hidden_states=outputs.hidden_states,
+        )
```
Comment (sgugger) [inline src/transformers/models/cvt/modeling_cvt.py:753]:
```suggestion
        return ImageClassifierOutput(loss=loss, logits=logits, hidden_states=outputs.hidden_states)
```

**Review by sgugger — APPROVED** [review]
Thanks for adding this new model! On top of the comments, please wait for #17304 to be merged and then rebase and run `make fix-copies` to fix the dummy objects (not added in this PR as the util is broken right now).

File: src/transformers/models/cvt/configuration_cvt.py [hunk 4]
```diff
@@ -0,0 +1,147 @@
...
+            The number of input channels.
+        patch_sizes (`List[int]`, *optional*, defaults to `[7, 3, 3]`):
+            The kernel size of each encoder's patch embedding.
+        patch_stride (`List[int]`, *optional*, defaults to `[4, 2, 2]`):
+            The stride size of each encoder's patch embedding.
+        patch_padding (`List[int]`, *optional*, defaults to `[2, 1, 1]`):
+            The padding size of each encoder's patch embedding.
+        embed_dim (`List[int]`, *optional*, defaults to `[64, 192, 384]`):
+            Dimension of each of the encoder blocks.
+        num_attention_heads (`List[int]`, *optional*, defaults to `[1, 3, 6]`):
```
Comment (LysandreJik) [inline src/transformers/models/cvt/configuration_cvt.py:50]:
```suggestion
        num_heads (`List[int]`, *optional*, defaults to `[1, 3, 6]`):
```
Renaming according to how the variable is currently named

**Review by LysandreJik — APPROVED** [review]
LGTM! Only left one nit on a variable that isn't named correctly.

Looks great, thanks a lot for working on it @AnugunjNaman, @NielsRogge!

**Pull request merged**